File: frontend/src/utils/campaignChannels.js
/** Orden de canales: preserva el orden en que el usuario los eligió / guardó. */
export const CHANNEL_ORDER = ['linkedin', 'email', 'whatsapp', 'call']

export const CHANNEL_LABELS = {
  linkedin: 'LinkedIn',
  email: 'Email',
  whatsapp: 'WhatsApp',
  call: 'Llamada',
}

/** Canales por defecto: LinkedIn → Email → WhatsApp → Llamada asistida. */
export const DEFAULT_ALLOWED_CHANNELS = ['linkedin', 'email', 'whatsapp', 'call']

/** Devuelve la lista en el orden dado (sin forzar LinkedIn→Email→WhatsApp). */
export function orderChannels(selected) {
  const out = []
  const seen = new Set()
  const allowed = new Set(CHANNEL_ORDER)
  for (const c of Array.isArray(selected) ? selected : []) {
    const lc = String(c).toLowerCase()
    if (allowed.has(lc) && !seen.has(lc)) {
      seen.add(lc)
      out.push(lc)
    }
  }
  return out
}

export function formatChannelsSummary(channels) {
  const ch = orderChannels(channels)
  if (!ch.length) {
    return 'Ninguno'
  }
  return ch.map((id) => CHANNEL_LABELS[id] ?? id).join(' → ')
}

File: backend/app/schemas/campaign_channels.py
from __future__ import annotations

import json

CHANNEL_PRIORITY = ("linkedin", "email", "whatsapp")

# Canales por defecto: LinkedIn + Email + WhatsApp.
DEFAULT_MVP_CHANNELS = ("linkedin", "email", "whatsapp")


def normalize_allowed_channels(channels: list[str] | None) -> list[str]:
    """Preserva el orden que eligió el usuario (no reordenar por prioridad fija)."""
    allowed = {"linkedin", "email", "whatsapp"}
    if channels is None:
        return list(DEFAULT_MVP_CHANNELS)
    if not channels:
        raise ValueError("Debe habilitarse al menos un canal.")
    seen: set[str] = set()
    raw: list[str] = []
    for c in channels:
        lc = str(c).strip().lower()
        if lc == "call":
            # Retired assisted-call channel; ignore in stored campaign configs.
            continue
        if lc not in allowed:
            raise ValueError(f"Canal inválido: {c}")
        if lc not in seen:
            seen.add(lc)
            raw.append(lc)
    if not raw:
        raise ValueError("Debe habilitarse al menos un canal.")
    return raw


def coerce_allowed_channels(raw: object) -> list[str]:
    """Acepta lista JSON (modelo) o TEXT legacy en SQLite."""
    if raw is None:
        return list(DEFAULT_MVP_CHANNELS)
    data = raw
    if isinstance(data, str):
        try:
            data = json.loads(data)
        except json.JSONDecodeError:
            return list(DEFAULT_MVP_CHANNELS)
    if isinstance(data, list):
        try:
            return normalize_allowed_channels([str(x).lower().strip() for x in data])
        except ValueError:
            return list(DEFAULT_MVP_CHANNELS)
    return list(DEFAULT_MVP_CHANNELS)
